[RPC] Transaction details in getblock

achow101 opened the pull request:
Adds an optional parameter extraVerbose to getblock to have transaction details displayed in a getblock call.

dcousens commented:
NACK,  is two RPC calls really so bad?

achow101 commented:
@dcousens yes, two RPC calls is really bad, especially when you want the details of all the transactions in a block. Then you end up running getrawtransaction thousands of times. Also, AFAICT, this doesn't require the txindex in order to get the transactions, unlike getrawtransaction.

dcousens commented:
`getrawtransaction`?
In a batched RPC call,  the overhead is literally just 1RTT more.
Not N,  1.

achow101 commented:
> In a batched RPC call, the overhead is literally just 1RTT more.
Not N (aka, thousands), 1.

How?

> edit: In fact, adding that option would be significantly more useful than this RTT optimisation IMHO.

Well the functionality to this was already in the code. The BlockToJson method took a parameter for txDetails which was by default false. This just lets you set that to true.

dcousens commented:
http://www.jsonrpc.org/specification#batch

What language are you using? Your RPC library should be able to handle this quite easily.

achow101 commented:
Huh. Didn't know that. I've been using bash with either curl or bitcoin-cli and python.

laanwj commented:
Saving roundtrip time is not a valid reason to add a RPC.
>  Also, AFAICT, this doesn't require the txindex in order to get the transactions, unlike getrawtransaction.

This is the only important fact here: you *cannot* get this information any other way. getrawtransaction won't get you transactions in blocks, at least without tx index.

dcousens commented:
@laanwj would it maybe be better to add functionality via getrawtransaction w/ a block id?

laanwj commented:
> would it maybe be better to add functionality via getrawtransaction w/ a block id? (perhaps in addition to this)

Well yes a "gather transaction [X,...] from prespecified block Y" RPC call could be useful in some rare cases. But on the other hand it'll still have to read the entire block from disk and deserialize it. It's terribly inefficient already.
So if it decoded the entire block why not report details for the entire block?

dcousens commented:
@laanwj indeed!  Forgot about the whole deserialization.

luke-jr commented:
Concept ACK for all the reasons @laanwj already discussed.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -689,11 +689,12 @@ UniValue getblockheader(const UniValue& params, bool fHelp)
+    if (fHelp || params.size() < 1 || params.size() > 3)
         throw runtime_error(
-            "getblock \"hash\" ( verbose )\n"
+            "getblock \"hash\" ( verbose ) ( extraVerbose )\n"
```
Double boolean here seems ugly. Maybe allow verbose to be boolean or a number 0-2 (with 2 being extraVerbose)?

laanwj replied:
It is ugly, but overloading on type in JSON which is essentially dynamically typed is also ugly.

Named parameters as implemented in #8811 would make this more bearable.

laanwj replied:
I do agree that for a new interface, a scale for verbosity would have made more sense instead of a boolean.

laanwj commented:
I'm starting to realize that @luke-jr's idea to accept 0-2 isn't such a bad idea after all. Just rename the argument  'verbosityLevel'. It is easier to use and understand from a user perspective than two booleans (we regularly get confused there in the tests ourselves). Also it'd remove having to deal with the redundant and nonsensical combo `verbose=false extraVerbose=true`.

For consistency we should do the same in `getrawtransaction`. This method currently accepts a `verbose` argument that can be `0` or `1` but not `false` or `true`. The mapping `false:0,true:1` should be added there (not in this pull though).

achow101 commented:
How should it document in the help message that the old behavior is still accepted?

achow101 commented:
Rebased and made verbose an int from 0-2

luke-jr commented:
@achow101 The old behaviour should be considered deprecated, and therefore not documented.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -691,13 +691,17 @@ UniValue getblock(const JSONRPCRequest& request)
             "1. \"hash\"          (string, required) The block hash\n"
-            "2. verbose           (boolean, optional, default=true) true for a json object, false for the hex encoded data\n"
-            "\nResult (for verbose = true):\n"
+            "2. verbose           (boolean, optional, default=1) 0 for hex encoded data, 1 for a json object, and 2 for json object with transaction data\n"
```
Should be renamed to verbosity.

luke-jr commented:
Also, fix the commit description...

achow101 commented:
@luke-jr done

achow101 commented:
merge please?

achow101 commented:
merge please?

luke-jr reviewed (approved):
utACK

luke-jr commented:
(Needs rebase)

luke-jr committed:
RPC: Allow multiple names for parameters

achow101 commented:
rebased

pstratem commented:
Is this compatible with callers of getblock which send true instead of 0/1 ?

achow101 commented:
@pstratem Yes. The old style of calling is maintained for compatibility.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -721,8 +724,72 @@ UniValue getblock(const JSONRPCRequest& request)
+            "  \"version\" : n,         (numeric) The block version\n"
+            "  \"versionHex\" : \"00000000\", (string) The block version formatted in hexadecimal\n"
+            "  \"merkleroot\" : \"xxxx\", (string) The merkle root\n"
+            "  \"tx\" : [               (array of Objects) The transactions\n"
```
I'd suggest not including this in the help output, and just having something like:

```c
           "  \"tx\" : [               (array of Objects) The transactions. Format is the same as for getrawtransaction call\n"
           "         ,...\n"
           "  ],\n"
```

If the output format of `TxToJSON()` ever changes, I'm sure that changing this help text will be forgotten!

sipa commented:
Needs rebase.

achow101 committed:
Use a verbosity instead of two verbose parameters

Verbose is changed to an int. This can have values from 0-2 for each level of verbosity.
Verbosity level 2 has transaction details displayed in the results.

achow101 commented:
rebased. also added @jnewbery's suggestion with the help text.

kallewoof reviewed (commented):
utACK b779f3093d7b68ee09305a5294c48825dce8ae1f

paveljanik commented on src/rpc/blockchain.cpp:
```diff
@@ -716,8 +719,29 @@ UniValue getblock(const JSONRPCRequest& request)
             "}\n"
-            "\nResult (for verbose=false):\n"
-            "\"data\"             (string) A string that is serialized, hex-encoded data for block 'hash'.\n"
+            "\nResult (for verbosity = 2):\n"
```
What about mentioning only the differences to verbosity=1 instead?

achow101 replied:
something like this?

```
{
..., Same as for verbosity=1
"tx" : [           (array of Objects) The transactions in the format of the getrawtransaction RPC.
         , ...
],
... Same as for verbosity=1
}
```

paveljanik replied:
Yup, or maybe even better:

```
Additional output (for verbosity=1):...
```

achow101 commented:
@paveljanik I made the change.

laanwj commented:
Tested ACK e3c9f2d
